src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs

```diff
@@ -70,102 +68,113 @@ private unsafe static long GetAsciiStringAsLong(string str)
                 {
-                    var input = block.Array;
-                    for (var i = 0; i < following; i++)
+                    int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+
+                    if (following > 0)
                     {
-                        output[i + outputOffset] = (char)input[i + inputOffset];
+                        fixed (byte* blockStart = block.Array)
+                        {
+                            var input = blockStart + inputOffset;
+                            var i = 0;
+                            while (i + 11 < following)
+                            {
+                                i += 12;
+                                *(output) = (char)*(input);
+                                *(output + 1) = (char)*(input + 1);
+                                *(output + 2) = (char)*(input + 2);
+                                *(output + 3) = (char)*(input + 3);
+                                *(output + 4) = (char)*(input + 4);
+                                *(output + 5) = (char)*(input + 5);
+                                *(output + 6) = (char)*(input + 6);
+                                *(output + 7) = (char)*(input + 7);
+                                *(output + 8) = (char)*(input + 8);
+                                *(output + 9) = (char)*(input + 9);
+                                *(output + 10) = (char)*(input + 10);
+                                *(output + 11) = (char)*(input + 11);
+                                output += 12;
+                                input += 12;
+                            }
+                            if (i + 6 < following)
+                            {
+                                i += 6;
+                                *(output) = (char)*(input);
+                                *(output + 1) = (char)*(input + 1);
+                                *(output + 2) = (char)*(input + 2);
+                                *(output + 3) = (char)*(input + 3);
+                                *(output + 4) = (char)*(input + 4);
+                                *(output + 5) = (char)*(input + 5);
+                                output += 6;
+                                input += 6;
+                            }
+                            if (i + 3 < following)
+                            {
+                                i += 4;
+                                *(output) = (char)*(input);
+                                *(output + 1) = (char)*(input + 1);
+                                *(output + 2) = (char)*(input + 2);
+                                *(output + 3) = (char)*(input + 3);
+                                output += 4;
+                                input += 4;
+                            }
+                            while (i < following)
+                            {
+                                i++;
+                                *(output++) = (char)*(input++);
```

Comment:
It may not be measurable with only 3 iterations but

    *output = (char)*input;
    output++;
    input++;

will be slightly faster than

    *(output++) = (char)*(input++);